discovery/sync_manager.go

```diff
@@ -0,0 +1,640 @@
+// sync with it.
+func (m *SyncManager) forceHistoricalSync() {
+	m.syncersMtx.Lock()
+	defer m.syncersMtx.Unlock()
+
+	candidatesChosen := make(map[routing.Vertex]struct{})
+	s := m.chooseRandomSyncer(candidatesChosen)
+	for s != nil {
+		candidatesChosen[s.cfg.peerPub] = struct{}{}
+		err := m.transitionSyncer(s, HistoricalSync)
+		if err == nil {
+			break
+		}
+
+		log.Errorf("Unable to transition GossipSyncer(%x) to %v: %v",
+			s.cfg.peerPub, HistoricalSync, err)
+
+		s = m.chooseRandomSyncer(candidatesChosen)
+	}
+}
+
+// chooseRandomSyncer returns a random non-active syncer that's eligible for a
+// sync transition. A blacklist can be used to skip any previously chosen
+// candidates.
+//
+// NOTE: It's possible for a nil value to be returned if there are no eligible
+// candidate syncers.
+//
+// NOTE: This method requires the syncersMtx read lock to be held.
+func (m *SyncManager) chooseRandomSyncer(
+	blacklist map[routing.Vertex]struct{}) *GossipSyncer {
+
+	for _, s := range m.inactiveSyncers {
+		if blacklist != nil {
+			if _, ok := blacklist[s.cfg.peerPub]; ok {
+				continue
+			}
+		}
+		if s.syncState() != chansSynced {
+			continue
+		}
+		return s
+	}
+	return nil
+}
+
+// InitSyncState is called by outside sub-systems when a connection is
+// established to a new peer that understands how to perform channel range
+// queries. We'll allocate a new gossip syncer for it, and start any goroutines
+// needed to handle new queries.
+//
+// TODO(wilmer): Only mark as ActiveSync if this isn't a channel peer.
+func (m *SyncManager) InitSyncState(peer lnpeer.Peer) {
+	m.syncersMtx.Lock()
+	defer m.syncersMtx.Unlock()
+
+	// If we already have a syncer, then we'll exit early as we don't want
+	// to override it.
+	nodeID := routing.Vertex(peer.PubKey())
+	if _, ok := m.gossipSyncer(nodeID); ok {
+		return
+	}
+
+	log.Infof("Creating new GossipSyncer for peer=%x", nodeID[:])
+
+	encoding := lnwire.EncodingSortedPlain
+	s := newGossipSyncer(GossipSyncerCfg{
+		chainHash:     m.cfg.ChainHash,
+		peerPub:       nodeID,
+		channelSeries: m.cfg.ChanSeries,
+		encodingType:  encoding,
+		chunkSize:     encodingTypeToChunkSize[encoding],
+		sendToPeer: func(msgs ...lnwire.Message) error {
+			return peer.SendMessage(false, msgs...)
+		},
+	})
+
+	// If we don't have any active syncers, we'll default to forcing a full
+	// sync with the first peer we connect to, to ensure we get as much of
+	// the graph as possible. This is mostly done for light clients, which
+	// shouldn't be receiving new updates anyway due to the high bandwidth
+	// consumption that stems from it.
+	if m.cfg.NumActiveSyncers == 0 && len(m.inactiveSyncers) == 0 {
+		s.syncType = uint32(HistoricalSync)
+		m.inactiveSyncers[nodeID] = s
+		s.Start()
+		return
+	}
+
+	// If we've yet to reach our desired number of active syncers, then
+	// we'll use this one.
+	if len(m.activeSyncers) < m.cfg.NumActiveSyncers {
+		s.syncType = uint32(ActiveSync)
```

Comment:
As mentioned, I think it is easier to just let this be `Passive`, then let the call to `enqueueActiveSyncer` transition it into `Active`.

Same for `HistoricalSync` above: let it start `Passive`, then signal to the `syncerHandler` to force a historical sync.

Reply:
The transition requires the syncer to be in a `chansSynced` state, so if we go with this approach they'll have to go through the state machine twice (once the first time they're started, another after the transition).